### PR TITLE
feat(sonata-provider): add sonata media private file provider

### PR DESCRIPTION
**Pull request by bmaziere**
I am targeting this branch, because this is the only one.

## Changelog

```markdown
### Added
- Added Sonata Media Private File Provider Service 
```

## To do
- [ ] Add tests

### CODE REVIEW

File: docs/30-sonata-media-private-file-provider.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+Sonata Media Private File Provider
+==================================
+
+This file provider extends the file provider of [Sonata Media package][1] and allow to store
```
Comment (mremi) [inline docs/30-sonata-media-private-file-provider.md]: `allows`

File: docs/30-sonata-media-private-file-provider.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+a public url.
+
+Configuration
+-------------
+
+```yaml
+sonata_helpers:
+    sonata_media_private_file_provider:
+        url_prefix: /admin/private #default
+        storage_path: data/media #default - will be concatenated with kernel.project_dir
```
Comment (mremi) [inline docs/30-sonata-media-private-file-provider.md]: I think you should add `%kernel.project_dir%/` to be able to configure something like `/data/foo/bar`

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        $configuration = new Configuration();
+        $config        = $this->processConfiguration($configuration, $configs);
+        $loader        = new YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
+        $loader->load('services.yaml');
+
+        $this->configureSonataMediaPrivateFileProvider($config['sonata_media_private_file_provider'], $container);
+    }
+
+    public function prepend(ContainerBuilder $container)
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: missing phpdoc

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        $bundles           = $container->getParameter('kernel.bundles');
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: unused

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        $configs           = $container->getExtensionConfig($this->getAlias());
+        $config            = $this->processConfiguration(new Configuration(), $configs);
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: looks weird

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            $config['sonata_media_private_file_provider']['allowed_extensions'] = $sonataMediaConfig[0]['providers']['file']['allowed_extensions'];
+        }
+        if (empty($config['sonata_media_private_file_provider']['allowed_mime_types'])) {
+            $config['sonata_media_private_file_provider']['allowed_mime_types'] = $sonataMediaConfig[0]['providers']['file']['allowed_mime_types'];
+        }
+
+        $container->prependExtensionConfig('sonata_helpers', $config);
+    }
+
+    private function configureSonataMediaPrivateFileProvider(array $config, ContainerBuilder $container): void
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: missing phpdoc

File: src/Resources/config/services.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+services:
```
Comment (mremi) [inline src/Resources/config/services.yaml]: why yaml?

Comment (bmaziere) [reply src/Resources/config/services.yaml]: why not? tired of xml. didn't find anywhere that it has to be done with xml no?

Comment (mremi) [reply src/Resources/config/services.yaml]: https://symfony.com/doc/current/bundles/configuration.html#supporting-xml

File: docs/30-sonata-media-private-file-provider.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+a public url.
+
+Configuration
+-------------
+
+```yaml
+sonata_helpers:
+    sonata_media_private_file_provider:
+        url_prefix: /admin/private #default
+        storage_path: %kernel.project_dir%/data/media #default
```
Comment (mremi) [inline docs/30-sonata-media-private-file-provider.md]: missing `'` over parameter

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+     * @param ContainerBuilder $container
+     */
+    private function configureSonataMediaPrivateFileProvider(array $config, ContainerBuilder $container): void
+    {
+        $container
+            ->findDefinition('sonata_helpers.private.cdn.server')
+            ->replaceArgument(0, $config['url_prefix']);
+        $container
+            ->findDefinition('sonata_helpers.private.adapter.local')
+            ->replaceArgument(0, sprintf('%s/%s', $container->getParameter('kernel.project_dir'),
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: No need anymore to append parameter

File: src/DependencyInjection/Configuration.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+     * {@inheritdoc}
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $rootNode    = $treeBuilder->root('sonata_helpers');
+
+        $rootNode
+            ->children()
+                ->arrayNode('sonata_media_private_file_provider')
```
Comment (qsomazzi) [inline src/DependencyInjection/Configuration.php]: would be nice to add tests for this no ?

Comment (bmaziere) [reply src/DependencyInjection/Configuration.php]: will add it later if possible

File: tests/DependencyInjection/SonataHelpersExtensionTest.php [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        $this->extension = new SonataHelpersExtension();
+    }
+
+    /**
+     * Assert the sonata media private file provider is well configured.
+     */
+    public function testSonataMediaConfig(): void
+    {
+        $container = $this->getMockBuilder(ContainerBuilder::class)->setMethods(['findDefinition'])->getMock();
```
Comment (mremi) [inline tests/DependencyInjection/SonataHelpersExtensionTest.php]: `createPartialMock` 😉

**Pull request merged**